src/raw/document_buf.rs

```diff
@@ -136,19 +159,169 @@ impl RawDocumentBuf {
     /// assert_eq!(doc.into_vec(), b"\x05\x00\x00\x00\x00".to_vec());
     /// # Ok::<(), Error>(())
     /// ```
-    pub fn into_vec(self) -> Vec<u8> {
+    pub fn into_bytes(self) -> Vec<u8> {
         self.data
     }
+
+    /// Append a key value pair to the end of the document without checking to see if
+    /// the key already exists.
+    ///
+    /// It is a user error to append the same key more than once to the same document, and it may
+    /// result in errors when communicating with MongoDB.
+    ///
+    /// If the provided key contains an interior null byte, this method will panic.
+    ///
+    /// ```
+    /// # use bson::raw::Error;
+    /// use bson::{doc, raw::RawDocumentBuf};
+    ///
+    /// let mut doc = RawDocumentBuf::new();
+    /// doc.append("a string", "some string");
+    /// doc.append("an integer", 12_i32);
+    ///
+    /// let mut subdoc = RawDocumentBuf::new();
+    /// subdoc.append("a key", true);
+    /// doc.append("a document", subdoc);
+    ///
+    /// let expected = doc! {
+    ///     "a string": "some string",
+    ///     "an integer": 12_i32,
+    ///     "a document": { "a key": true },
+    /// };
+    ///
+    /// assert_eq!(doc.to_document()?, expected);
+    /// # Ok::<(), Error>(())
+    /// ```
+    pub fn append(&mut self, key: impl Into<String>, value: impl Into<OwnedRawBson>) {
```

Comment:
Since this only appends to the end, we can't call it `insert` (which would imply key lookup). I opted for something different than `push` though which seems to be coupled with stacks/lists rather than key-value stores. Let me know if these seems right.